core/Pos/Binary/Crypto.hs

```diff
@@ -43,263 +34,192 @@ import           Pos.Crypto.Signing         (ProxyCert (..), ProxySecretKey (..)
-constantSizedBinaryToStoreGet :: Binary.Binary a => Int -> Store.Peek a
-constantSizedBinaryToStoreGet bytes = do
-    x <- getBytes bytes
-    case Binary.decodeOrFail (BSL.fromStrict x) of
-        Left (_, _, err) -> fail err
-        Right (bs, _, res)
-            | BSL.null bs -> pure res
-            | otherwise   -> fail "unconsumed input"
-
--- [CSL-1122] TODO: more efficient 'put' and 'get' are possible if 'Store'
--- instances are added into pvss-haskell
-#define BiPvss(T, PT, BYTES) \
+#define BiPvss(T, PT) \
   instance Bi T where {\
-    size = ConstSize BYTES ;\
-    put = labelP "T" . putBytes . BSL.toStrict . Binary.encode ;\
-    get = label "T" $ constantSizedBinaryToStoreGet BYTES };\
+    encode = encodeBinary ;\
+    decode = decodeBinary };\
   deriving instance Bi PT ;\
 
-BiPvss (Pvss.PublicKey, VssPublicKey, 33)    -- yes it's 33 and not 32
-BiPvss (Pvss.KeyPair, VssKeyPair, 65)        -- 32+33
-BiPvss (Pvss.Secret, Secret, 33)
-BiPvss (Pvss.DecryptedShare, Share, 101)     -- 4+33+64
-BiPvss (Pvss.EncryptedShare, EncShare, 101)
-BiPvss (Pvss.Proof, SecretProof, 64)
-
-instance Store.Store Pvss.ExtraGen where
-    size = ConstSize 33
-    poke = labelP "Pvss.ExtraGen" . putBytes . BSL.toStrict . Binary.encode
-    peek = label "Pvss.ExtraGen" $ constantSizedBinaryToStoreGet 33
-instance Store.Store Pvss.Commitment where
-    size = ConstSize 33
-    poke = labelP "Pvss.Commitment" . putBytes . BSL.toStrict . Binary.encode
-    peek = label "Pvss.Commitment" $ constantSizedBinaryToStoreGet 33
-
-Store.makeStore ''SecretSharingExtra
+BiPvss (Pvss.PublicKey, VssPublicKey)
+BiPvss (Pvss.KeyPair, VssKeyPair)
+BiPvss (Pvss.Secret, Secret)
+BiPvss (Pvss.DecryptedShare, Share)
+BiPvss (Pvss.EncryptedShare, EncShare)
+BiPvss (Pvss.Proof, SecretProof)
+
+instance Bi Pvss.ExtraGen where
+    encode = encodeBinary
+    decode = decodeBinary
+
+instance Bi Pvss.Commitment where
+    encode = encodeBinary
+    decode = decodeBinary
+
 instance Bi SecretSharingExtra where
-    put = labelP "SecretSharingExtra" . Store.poke
-    get = label "SecretSharingExtra" $ Store.peek
-    size = Store.size
+    encode (SecretSharingExtra eg comms) = encodeListLen 2
+                                        <> encode eg
+                                        <> encode comms
+    decode = SecretSharingExtra
+          <$  enforceSize "SecretSharingExtra" 2
+          <*> decode
+          <*> decode
 
 deriving instance Bi (AsBinary SecretSharingExtra)
 
 ----------------------------------------------------------------------------
 -- SecretSharing AsBinary
 ----------------------------------------------------------------------------
 
-#define BiMacro(B, BYTES) \
+#define BiMacro(B) \
   instance Bi (AsBinary B) where {\
-    size = ConstSize BYTES ;\
-    put (AsBinary bs) = putBytes bs ;\
-    get = label "B (BYTES bytes)" $ AsBinary <$> getBytes BYTES}; \
+    encode (AsBinary bs) = encode bs ;\
+    decode = AsBinary <$> decode}; \
```

Comment:
@gromakovsky My 2 cents: these might be all super valid concerns, but I feel they somewhat fall in the "to-be-addressed-later-to-this-PR" bucket. Given my limited knowledge of the codebase, I have limited myself to change only the things I needed to change, not knowing all the backstory about certain choices (as you have explained to me). The TL;DR is that we should first merge this _and then_ tackle those, or the risk of never getting this merged is real 😉  What do you think?

Generally speaking, to reassure you on the whole `AsBinary` story; it's generally true that CBOR is a var-len encoding, but in the case of `AsBinary` we tend to call `fromBinary` and `asBinary` on things of known size, like Public Keys etc. So, if the input string is fixed, also the overhead CBOR introduces is predictable, so I think for now everything should continue working, but of course it might make sense to scrap the `checkLen` call as you pointed out or review how we are doing things in this area now that we use CBOR.